src/lsm/tree_fuzz.zig

```diff
@@ -599,96 +601,152 @@ fn EnvironmentType(comptime table_usage: TableUsage) type {
-            // all keys found by the scan must exist in our model.
-            var tree_value_last: ?Value = null;
-            for (tree_values) |tree_value| {
+            const model_values = model.scan(
+                scan_range.min,
+                scan_range.max,
+                scan_range.direction,
+                env.scan_results_model,
+            );
 
-                // Asserting boundaries.
+            // Unlike the model, the tree can return some amount of tombstones in the result set.
+            // They must be filtered out before comparison!
+            var tombstone_count: usize = 0;
+            for (tree_values, 0..) |tree_value, index| {
                 assert(scan_range.min <= Table.key_from_value(&tree_value));
                 assert(Table.key_from_value(&tree_value) <= scan_range.max);
-
-                // Asserting direction.
-                if (tree_value_last) |value_last| {
-                    switch (scan_range.direction) {
-                        .ascending => assert(Table.key_from_value(&tree_value) >
-                            Table.key_from_value(&value_last)),
-                        .descending => assert(Table.key_from_value(&tree_value) <
-                            Table.key_from_value(&value_last)),
-                    }
-                }
-                tree_value_last = tree_value;
-
-                // Compare result to model.
-                if (model.get(Table.key_from_value(&tree_value))) |model_value| {
-                    switch (table_usage) {
-                        .general => {
-                            assert(std.mem.eql(
-                                u8,
-                                std.mem.asBytes(&model_value),
-                                std.mem.asBytes(&tree_value),
-                            ));
-                        },
-                        .secondary_index => {
-                            // secondary_index only preserves keys - may return old values
-                            assert(Table.key_from_value(&model_value) ==
-                                Table.key_from_value(&tree_value));
-                        },
-                    }
+                if (Table.tombstone(&tree_value)) {
+                    tombstone_count += 1;
                 } else {
-                    assert(Table.tombstone(&tree_value));
+                    if (tombstone_count > 0) {
+                        tree_values[index - tombstone_count] = tree_value;
+                    }
                 }
             }
 
-            // Asserting the negative space:
-            // All keys existing in our model must be checked against the scan range.
-            if (scan_range.direction == .descending) std.mem.reverse(
+            try std.testing.expectEqualSlices(
                 Value,
-                tree_values,
+                tree_values[0 .. tree_values.len - tombstone_count],
+                model_values[0 .. tree_values.len - tombstone_count],
```

Comment:
Checking my understanding: Are we are deliberately ignoring any extra values from `model_values`, because they didn't fit in the tree's scan due to tombstones?

Reply:
That was the idea, but the implementation was incorrect (it wrongly trusted the length of tree_values). Fixed the impl and added an assert that would have caught that bug.